Sources/Confidence/ConfidenceValue.swift

```diff
@@ -0,0 +1,272 @@
+    }
+
+    public init(null: ()) {
+        self.value = .null
+    }
+
+    private init(valueInternal: ConfidenceValueInternal) {
+        self.value = valueInternal
+    }
+
+    public func asBoolean() -> Bool? {
+        if case let .boolean(bool) = value {
+            return bool
+        }
+
+        return nil
+    }
+
+    public func asString() -> String? {
+        if case let .string(string) = value {
+            return string
+        }
+
+        return nil
+    }
+
+    public func asInteger() -> Int64? {
+        if case let .integer(int64) = value {
+            return int64
+        }
+
+        return nil
+    }
+
+    public func asDouble() -> Double? {
+        if case let .double(double) = value {
+            return double
+        }
+
+        return nil
+    }
+
+    public func asDateComponents() -> DateComponents? {
+        if case let .date(dateComponents) = value {
+            return dateComponents
+        }
+
+        return nil
+    }
+
+    public func asDate() -> Date? {
+        if case let .timestamp(date) = value {
+            return date
+        }
+
+        return nil
+    }
+
+    public func asList() -> [ConfidenceValue]? {
+        if case let .list(values) = value {
+            return values.map { i in ConfidenceValue(valueInternal: i) }
+        }
+
+        return nil
+    }
+
+    public func asStructure() -> [String: ConfidenceValue]? {
+        if case let .structure(values) = value {
+            return values.mapValues { ConfidenceValue(valueInternal: $0) }
+        }
+
+        return nil
+    }
+
+    public func isNull() -> Bool {
+        if case .null = value {
+            return true
+        }
+
+        return false
+    }
+
+    public func type() -> ConfidenceValueType {
+        switch value {
+        case .boolean:
+            return .boolean
+        case .string:
+            return .string
+        case .integer:
+            return .integer
+        case .double:
+            return .double
+        case .date:
+            return .date
+        case .timestamp:
+            return .timestamp
+        case .list:
+            return .list
+        case .structure:
+            return .structure
+        case .null:
+            return .null
+        }
+    }
+
+    public static func == (lhs: ConfidenceValue, rhs: ConfidenceValue) -> Bool {
+        lhs.value == rhs.value
+    }
+}
+
+extension ConfidenceValue {
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        try container.encode(value)
+    }
+}
+
+public enum ConfidenceValueType: CaseIterable {
+    case boolean
+    case string
+    case integer
+    case double
+    case date
+    case timestamp
+    case list
+    case structure
+    case null
+}
+
+
+/// Serializable data structure meant for event sending via Confidence
+private enum ConfidenceValueInternal: Equatable, Encodable {
```

Comment:
Would it be more idiomatic to just make this the ConfidenceValue? I.e. Swift has a lot of great support for working with these (algebraic) enums. Do we provide anything by wrapping it, or are we just making it more cumbersome for Swift devs? Not sure about this, just want to hear others thoughts.

Reply:
The problem here was to keep algebraic enums and still keep control over the constructors to prevent, for example, the instantiation of heterogenous-type Lists. Maybe there are fancy ways to achieve this with pure enums, but I am unsure what value this brings to the final users. As long as we keep the enums internally to work with exhaustive switch-case code, I think we might have found a good sweet spot

Reply:
Proposal is to merge this as is and re-evaluate these protocols and APIs during a QA session pre-release